graph/formats/rdf/iso_canonical.go

```diff
@@ -0,0 +1,953 @@
+		return false
+	}
+	return unquoteIRI(si.Label.Value) < unquoteIRI(sj.Label.Value)
+}
+func (s c14nStatements) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+
+// hashBNodes returns the hashed blank nodes of the graph described by statements
+// using the provided hash function. Hashes are initialised with zero.
+//
+// This is algorithm 1 in doi:10.1145/3068333.
+func hashBNodes(statements []*Statement, h hash.Hash, zero []byte, hash0 map[string][]byte) (hash table, disjoint bool) {
+	curr := newTable()
+	for _, s := range statements {
+		for i, t := range []string{
+			s.Subject.Value,
+			s.Predicate.Value,
+			s.Object.Value,
+			s.Label.Value,
+		} {
+			switch {
+			case i == 3 && t == "":
+				continue
+			case strings.HasPrefix(t, "_:"):
+				if hash0 == nil {
+					curr.set(t, zero)
+				} else {
+					curr.set(t, hash0[t])
+				}
+			case strings.HasPrefix(t, "<") && strings.HasSuffix(t, ">"):
+				h.Reset()
+				h.Write([]byte(t[1 : len(t)-1])) //nolint:errcheck
+				curr.set(t, h.Sum(nil))
+			default:
+				h.Reset()
+				h.Write([]byte(t)) //nolint:errcheck
+				curr.set(t, h.Sum(nil))
+			}
+		}
+	}
+
+	bag := newHashBag(h, curr)
+	last := curr
+	for {
+		curr, last = last.clone(), curr
+		for _, s := range statements {
+			if strings.HasPrefix(s.Subject.Value, "_:") {
+				var lab []byte
+				if s.Label.Value != "" {
+					lab = last.hashOf[s.Label.Value]
+				}
+				c := hashTuple(h, last.hashOf[s.Object.Value], last.hashOf[s.Predicate.Value], lab, []byte{'+'})
+				bag.add(s.Subject.Value, c)
+			}
+
+			if strings.HasPrefix(s.Object.Value, "_:") {
+				var lab []byte
+				if s.Label.Value != "" {
+					lab = last.hashOf[s.Label.Value]
+				}
+				c := hashTuple(h, last.hashOf[s.Subject.Value], last.hashOf[s.Predicate.Value], lab, []byte{'-'})
+				bag.add(s.Object.Value, c)
+			}
+
+			// This and the lab value above implement the label hashing
+			// required for RDF dataset hashing as described in
+			// https://doi.org/10.5281/zenodo.3154322 v1.0
+			// Readme.md#adaptation-of-the-algorithms-to-handle-datasets.
+			if strings.HasPrefix(s.Label.Value, "_:") {
+				c := hashTuple(h, last.hashOf[s.Subject.Value], last.hashOf[s.Predicate.Value], last.hashOf[s.Object.Value], []byte{'.'})
+				bag.add(s.Label.Value, c)
+			}
+		}
+
+		for t := range bag.hashesFor {
+			curr.set(t, bag.sum(t))
+		}
+
+		disjoint = curr.allUnique()
+		if disjoint || !curr.changedFrom(last) {
+			return curr, disjoint
+		}
+	}
+}
+
+// table is a collision aware hash collection for RDF terms.
+type table struct {
+	// hashOf holds the hash for each term.
+	hashOf map[string][]byte
+	// termsFor holds the set of nodes in
+	// the second key for terms that share
+	// the hash in the first key.
+	termsFor map[string]map[string]bool
+
+	// isBlank and blanks are the set of blank
+	// nodes.
+	// isBlank is nil for cloned tables.
+	isBlank map[string]bool
+	// blanks is nil for tables created
+	// with newTable.
+	blanks []string
```

Comment:
I'd rather not if it can be avoided. Is there a reason you would like to see this?

Reply:
Currently you would have to rely on the above comments to know that `isBlank` and `blanks` carry this additional meaning. I thought maybe this could lead to surprises in the future?

Reply:
Having a separate field for each of these leaves you in the same situation; the information is stored in two places and they can disagree. An alternative approach would be to provide methods for this;

```
// wasCloned returns whether t is a parent or child of a cloning operation.
func (t *table) wasCloned() bool { return t.isBlank == nil }

// isNew returns whether t is a new table.
func (t *table) isNew() bool { return t.blanks == nil }
```

These document the meaning in a tied manner and cost nothing. Does that sound OK to you?

```diff
index 3306d37e..7f66ce1b 100644
--- a/graph/formats/rdf/iso_canonical.go
+++ b/graph/formats/rdf/iso_canonical.go
@@ -364,6 +364,12 @@ func newTable() *table {
        }
 }
 
+// wasCloned returns whether t is a parent or child of a cloning operation.
+func (t *table) wasCloned() bool { return t.isBlank == nil }
+
+// isNew returns whether t is a new table.
+func (t *table) isNew() bool { return t.blanks == nil }
+
 // clone returns a clone of the receiver.
 func (t *table) clone() *table {
        new := &table{
@@ -383,7 +389,7 @@ func (t *table) clone() *table {
                }
                new.termsFor[hash] = terms
        }
-       if t.blanks == nil {
+       if t.isNew() {
                t.blanks = make([]string, len(t.isBlank))
                i := 0
                for n := range t.isBlank {
@@ -424,7 +430,7 @@ func (t *table) set(term string, hash []byte) {
                t.termsFor[string(hash)] = map[string]bool{term: true}
        }
 
-       if t.isBlank != nil && strings.HasPrefix(term, "_:") {
+       if !t.wasCloned() && strings.HasPrefix(term, "_:") {
                // We are in the original table, so note
                // any blank node label that we see.
                t.isBlank[term] = true
@@ -434,7 +440,7 @@ func (t *table) set(term string, hash []byte) {
 // allUnique returns whether every term has an unique hash. allUnique
 // can only be called on a table that was returned by clone.
 func (t *table) allUnique() bool {
-       if t.blanks == nil {
+       if t.isNew() {
                panic("checked hash bag from uncloned table")
        }
        for _, term := range t.blanks {
@@ -448,7 +454,7 @@ func (t *table) allUnique() bool {
 // changedFrom returns whether the receiver has been updated from last.
 // changedFrom can only be called on a table that was returned by clone.
 func (t *table) changedFrom(last *table) bool {
-       if t.blanks == nil {
+       if t.isNew() {
                panic("checked hash bag from uncloned table")
        }
        for i, x := range t.blanks {
@@ -473,7 +479,7 @@ type hashBag struct {
 // the given hash table. newHashBag can only take a table parameter that
 // was returned by newTable.
 func newHashBag(h hash.Hash, t *table) hashBag {
-       if t.isBlank == nil {
+       if t.wasCloned() {
                panic("made hash bag from cloned table")
        }
        b := hashBag{hash: h, hashesFor: make(map[string][][]byte, len(t.isBlank))}
```

Reply:
Yes, that looks good.